pkg/config/msp_test.go

```diff
@@ -94,6 +108,799 @@ func TestMSPToProtoFailure(t *testing.T) {
+												},
+												"public_signer": "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCi0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K"
+											},
+											"tls_intermediate_certs": [],
+											"tls_root_certs": []
+										},
+										"type": 0
+									},
+									"version": "0"
+								}
+							},
+							"version": "0"
+						}
+					},
+					"mod_policy": "Admins",
+					"policies": {
+						"Admins": {
+							"mod_policy": "Admins",
+							"policy": {
+								"type": 3,
+								"value": {
+									"rule": "MAJORITY",
+									"sub_policy": "Admins"
+								}
+							},
+							"version": "0"
+						},
+						"Readers": {
+							"mod_policy": "Admins",
+							"policy": {
+								"type": 3,
+								"value": {
+									"rule": "ANY",
+									"sub_policy": "Readers"
+								}
+							},
+							"version": "0"
+						},
+						"Writers": {
+							"mod_policy": "Admins",
+							"policy": {
+								"type": 3,
+								"value": {
+									"rule": "ANY",
+									"sub_policy": "Writers"
+								}
+							},
+							"version": "0"
+						}
+					},
+					"values": {
+						"ACLs": {
+							"mod_policy": "Admins",
+							"value": {
+								"acls": {
+									"acl1": {
+										"policy_ref": "hi"
+									}
+								}
+							},
+							"version": "0"
+						},
+						"Capabilities": {
+							"mod_policy": "Admins",
+							"value": {
+								"capabilities": {
+									"V1_3": {}
+								}
+							},
+							"version": "0"
+						}
+					},
+					"version": "0"
+				}
+			},
+			"mod_policy": "",
+			"policies": {},
+			"values": {},
+			"version": "0"
+		},
+		"sequence": "0"
+	}
+	`
+
+	expectedConfigProto := &cb.Config{}
+	err = protolator.DeepUnmarshalJSON(bytes.NewBufferString(expectedConfig), expectedConfigProto)
+	gt.Expect(err).NotTo(HaveOccurred())
+
+	gt.Expect(config).To(Equal(expectedConfigProto))
+}
+
+func TestAddRootCAToMSPFailure(t *testing.T) {
+	t.Parallel()
+	gt := NewGomegaWithT(t)
+
+	channelGroup, err := baseApplicationChannelGroup()
+	gt.Expect(err).ToNot(HaveOccurred())
+	config := &cb.Config{
+		ChannelGroup: channelGroup,
+	}
+
+	tests := []struct {
+		spec        string
+		cert        *x509.Certificate
+		expectedErr string
+	}{
+		{
+			spec: "invalid key usage",
+			cert: &x509.Certificate{
+				KeyUsage: x509.KeyUsageKeyAgreement,
+			},
+			expectedErr: "certificate KeyUsage must be x509.KeyUsageCertSign",
+		},
+		{
+			spec: "certificate is not a CA",
+			cert: &x509.Certificate{
+				IsCA:     false,
+				KeyUsage: x509.KeyUsageCertSign,
+			},
+			expectedErr: "certificate must be a CA certificate",
+		},
+	}
+
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.spec, func(t *testing.T) {
+			gt := NewGomegaWithT(t)
+			err = AddRootCAToMSP(tc.cert, "Org1", config)
+			gt.Expect(err).To(MatchError(tc.expectedErr))
+		})
+	}
+}
+
+// creates a CRL with an empty list of revoked certificates
+// and returns it as ASN.1 encoded bytes
+func createCRLBytesForTest() []byte {
```

Comment:
This all looks very wrong to me. Copying unexported bits of the standard library is usually a hint that something is amiss.

1. Why aren’t we generating a real certificate and key?
2. Once we have the actual key, we can issue certificates and then revoke them using the “official” [CreateCRL](https://golang.org/pkg/crypto/x509/#Certificate.CreateCRL) method.

Reply:
Done